### PR TITLE
add process and thread addresses

### PR DESCRIPTION
**Pull request by yelhamer**
This PR adds addresses for processes and threads, which is a prerequisite for adding dynamic feature extraction to main.

TaskList:
- [x] add ProcessAddress
- [x] add ThreadAddress
- [x] update the Cape extractor
- [x] update the NullExtractors
- [x] update Feature freeze
- [x] update tests
- [x] fix mypy issues
- [x] decide and implement the file header #1614 
- [ ] add tests for `test_dynamic_freeze.py`

### Checklist

- [ ] No CHANGELOG update needed

- [ ] No new tests needed

- [ ] No documentation update needed

### CODE REVIEW

File: capa/features/address.py [hunk 1]
```diff
@@ -36,6 +36,58 @@ def __hash__(self):
...
+            f"ppid: {self.ppid}, " if self.ppid > 0 else "",
+            f"pid: {self.pid}",
+        )
+
+    def __hash__(self):
+        return hash((self.ppid, self.pid))
+
+    def __eq__(self, other):
+        assert isinstance(other, ProcessAddress)
+        if self.ppid > 0:
```
Comment (williballenthin) [inline capa/features/address.py:59]: im afraid this check here might lead to some surprising behavior, if one has ppid set and the other does not. are you sure this cannot be simplified to `return (self.pid, self.ppid) == (other.pid, other.ppid)`?

File: capa/features/address.py [hunk 2]
```diff
@@ -36,6 +36,58 @@ def __hash__(self):
...
+        return (self.ppid, self.pid) < (other.ppid, other.pid)
+
+
+class ThreadAddress(Address):
+    """addresses a thread in a dynamic execution trace"""
+
+    def __init__(self, process: ProcessAddress, tid: int):
+        assert tid >= 0
+        self.ppid = process.ppid
+        self.pid = process.pid
```
Comment (williballenthin) [inline capa/features/address.py:74]:
```suggestion
        self.process = process
```

Comment (williballenthin) [reply capa/features/address.py:74]: unless its important that these fields are broken out?

File: capa/features/extractors/cape/file.py [hunk 3]
```diff
@@ -24,8 +24,10 @@ def get_processes(static: Dict) -> Iterator[ProcessHandle]:
     """
 
     def rec(process):
-        inner: Dict[str, str] = {"name": process["name"], "ppid": process["parent_id"]}
-        yield ProcessHandle(pid=process["pid"], inner=inner)
+        address: ProcessAddress = ProcessAddress(pid=process["pid"], ppid=process["parent_id"])
+        inner: Dict[str, str] = {"name": process["name"]}
+        print(address)
```
Comment (williballenthin) [inline capa/features/extractors/cape/file.py:29]:
```suggestion
```

File: capa/features/extractors/null.py [hunk 4]
```diff
@@ -70,3 +78,49 @@ def get_instructions(self, f, bb):
...
+        for addr, feature in self.processes[p.address].features:
+            yield feature, addr
+
+    def get_threads(self, p):
+        for address in sorted(self.processes[p].threads.keys()):
+            yield ThreadHandle(address=address, inner={}, tid=address.pid)
+
+    def extract_thread_features(self, p, t):
+        for addr, feature in self.processes[p.address].threads[t.address].features:
+            yield feature, addr
```
Comment (williballenthin) [inline capa/features/extractors/null.py:123]: 👍🏼

File: capa/features/freeze/__init__.py [hunk 5]
```diff
@@ -135,6 +154,36 @@ class FileFeature(HashableModel):
...
 
 
+class ProcessFeature(HashableModel):
+    """
+    args:
+        process: the address of the process to which this feature belongs.
+        address: the address at which this feature is found.
+
+    process != address because, e.g., the feature may be found *within* the scope (process).
+    versus right at its starting address.
```
Comment (williballenthin) [inline capa/features/freeze/__init__.py:164]: i don't understand this - doesn't the `address` contain the information needed to pick the right process? can you maybe include an example that shows how these are not the same?

File: capa/features/freeze/__init__.py [hunk 6]
```diff
@@ -364,24 +515,68 @@ def loads(s: str) -> StaticFeatureExtractor:
...
+                    )
+                    for t in p.threads
+                },
+            )
+            for p in freeze.features.processes
+        },
+    )
+
+
+MAGIC = "capa000".encode("ascii")
```
Comment (williballenthin) [inline capa/features/freeze/__init__.py:543]: this is ok for now, but we may want to revert this before merge once we've discussed how to extend the file format appropriately.

**Review by williballenthin — APPROVED** [review]
nice work!

please respond to the inline comments and let me know when we should merge.

File: capa/features/freeze/__init__.py (hunk 5 above)
Comment (yelhamer) [reply capa/features/freeze/__init__.py:170]: isn't `address` the address returned by a feature? if so, the the environment strings extracted return a `NO_ADDRESS` for the moment (and so should the future registry and filename features I believe).
If I'm correct, then should we make the address of those features the same as the process address?

Comment (williballenthin) [reply capa/features/freeze/__init__.py:170]: yes, each feature is associated with an address.

when possible, we should associate features with the most specific possible address, like an API call feature associated with a DynamicAnalysisEventAddress (or similar). we should use ProcessAddress whenever the feature can only be related to a process but we're not sure at what specific point. an environment string is a good example of this - the string is related to some process, not a specific event. 

we should use `NO_ADDRESS` when we really have no relation to any specific object. im not sure if there are any good examples of this yet for dynamic analysis. perhaps CAPE version (but this isn't a feature...).

Comment (williballenthin) [reply capa/features/freeze/__init__.py:170]: > versus right at its starting address.

to clarify, i dont think we should consider ProcessAddress to refer to a specific point in time, like the event that spawns a process or the moment it springs into existence; rather, it should represent the entire Process over time.

**Comment by yelhamer** [comment]
I am not sure how to run the git actions tests without merging. the conflicts seem to resolve automatically however:

image

Regardless, the tests succeed on my machine:

image

**Comment by williballenthin** [comment]
@yelhamer please address merge conflicts then lets merge

**Comment by yelhamer** [comment]
@williballenthin I addressed the conflicts and updated the tests, it should be ready for merge now. If there are any other remarks please let me know and I'll work on them.

**Review by williballenthin — APPROVED** [review]
lgtm

**Pull request merged**